include/aws/io/dns.h

```diff
@@ -0,0 +1,602 @@
+#ifndef AWS_IO_DNS_H
+#define AWS_IO_DNS_H
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/io/io.h>
+
+#include <aws/common/atomics.h>
+
+struct aws_array_list;
+struct aws_event_loop_group;
+struct aws_string;
+
+/* use async destruction from the start */
+typedef void(aws_dns_on_destroy_completed_fn)(void *user_data);
+
+/*
+ * Types and discussion related to a new public interface for dns and host name resolution within the CRT.
+ *
+ * Why implement our own DNS:
+ *
+ *  (1) no other way of doing truly async DNS across all platforms (without spawning threads per resolve)
+ *  (2) make it possible to support, across all platforms and runtimes, DNS resolution that goes beyond what is
+ *  provided by default on a given platform (dns-over-https/tls, dns security extensions, etc...).
+ *  (3) makes it possible to support name resolution strategies that improve our ability to utilize AWS services
+ *
+ */
+
+/*
+ * Vocabulary/Concepts
+ *
+ * aws_dns_resolver - A thin interface around remote DNS queries.  Includes a default implementation that supports
+ * recursive and iterative queries.  Supports DNS over UDP.  (Eventually) Supports DNS over https and tls.  No support
+ * for non-internet-class queries.  Uses a vtable and can be replaced with a custom implementation.
+ *
+ * aws_dns_service_provider_config - Controls what host/port/protocol combos the default aws_dns_resolver uses to
+ * resolve queries.
+ *
+ * aws_host_resolution_service - A heavy-weight host name resolution service that contains CRT-specific and
```

Comment:
why not just call this aws_dns_caching_resolver or something? 

or aws_dns_cache